pyscf/tdscf/uks_sf.py

```diff
@@ -0,0 +1,220 @@
+        
+        if self.extype==0:
+            hdiag = numpy.hstack((e_ia_b2a.ravel(), -e_ia_a2b.ravel()))
+        elif self.extype==1:
+            hdiag = numpy.hstack((e_ia_a2b.ravel(), -e_ia_b2a.ravel()))
+
+        vresp = mf.gen_response_sf(mo_coeff, mo_occ, hermi=0,extype=extype,\
+                                   collinear_samples=collinear_samples)
+            
+        def vind(zs):
+            nz = len(zs)
+            zs = numpy.asarray(zs).reshape(nz,-1)
+            if self.extype==0:
+                zsb2a = (zs[:,:noccb*nvira]).reshape(nz,noccb,nvira)
+                zsa2b = (zs[:,noccb*nvira:]).reshape(nz,nocca,nvirb)
+            elif self.extype==1:
+                zsb2a = (zs[:,nocca*nvirb:]).reshape(nz,noccb,nvira)
+                zsa2b = (zs[:,:nocca*nvirb]).reshape(nz,nocca,nvirb)
+                
+            dmsb2a = lib.einsum('xov,po,qv->xpq', zsb2a, orbob, orbva.conj())
+            dmsa2b = lib.einsum('xov,po,qv->xpq', zsa2b, orboa, orbvb.conj())
+
+            v1aoA_b2a,v1aoA_a2b,v1aoB_b2a,v1aoB_a2b = vresp(numpy.asarray((dmsb2a,dmsa2b)))
+            v1A_b2a = lib.einsum('xpq,po,qv->xov', v1aoA_b2a, orbob, orbva.conj())
+            v1A_a2b = lib.einsum('xpq,po,qv->xov', v1aoA_a2b, orboa, orbvb.conj())
+            v1B_b2a = lib.einsum('xpq,po,qv->xov', v1aoB_b2a, orbob, orbva.conj())
+            v1B_a2b = lib.einsum('xpq,po,qv->xov', v1aoB_a2b, orboa, orbvb.conj())
+
+            # add the orbital energy difference in A matrix.
+            v1A_b2a += lib.einsum('ov,xov->xov', e_ia_b2a, zsb2a)
+            v1A_a2b += lib.einsum('ov,xov->xov', e_ia_a2b, zsa2b)
+            
+            if self.extype==0:
+                v1_top = v1A_b2a + v1B_b2a
+                v1_bom = - v1B_a2b - v1A_a2b
+            elif self.extype==1:
+                v1_top = v1A_a2b + v1B_a2b
+                v1_bom =-v1B_b2a - v1A_b2a
+                
+            hx = numpy.hstack((v1_top.reshape(nz,-1), v1_bom.reshape(nz,-1)))
+            return hx
+            
+        return vind, hdiag
+
+    def kernel(self, x0=None, nstates=None,extype=None,collinear_samples=None):
+        '''SF-TDDFT diagonalization solver
+        '''
+        cpu0 = (logger.process_clock(), logger.perf_counter())
+        mf = self._scf
+
+        self.check_sanity()
+        self.dump_flags()
+        
+        if extype is None:
+            extype = self.extype
+        else:
+            self.extype = extype
+        
+        if nstates is None:
+            nstates = self.nstates
+        else:
+            self.nstates = nstates
+            
+        if collinear_samples is None:
+            collinear_samples = self.collinear_samples
+        else:
+            self.nstates = nstates
+
+        log = logger.Logger(self.stdout, self.verbose)
+
+        vind, hdiag = self.gen_vind(self._scf,extype=extype,collinear_samples=collinear_samples)
+        precond = hdiag
+
+        def pickeig(w, v, nroots, envs):
```

Comment:
Remove this function, as it is not used in davidson solver.

Reply:
OK. This function will be removed.